File: src/plaude/ble/transfer.py
"""File download over BLE — voice packet capture approach."""

import struct
import time

from .client import PlaudClient
from .protocol import (
    CMD_SYNC_FILE_START, CMD_SYNC_FILE_TAIL,
    CMD_FILE_CHECKSUM, CMD_FILE_CHECKSUM_RSP,
    crc16_ccitt,
)


class DownloadError(Exception):
    """Raised when a file download fails deterministically."""


async def download_file(
    client: PlaudClient,
    session_id: int,
    file_size: int,
    verbose: bool = False,
) -> bytes:
    """Download a recording from the device.

    The device sends file data as proto_type=0x02 (voice) packets after
    SYNC_FILE_START, then sends SYNC_FILE_TAIL with CRC when done.

    Returns raw Opus bytes (with 9-byte per-frame headers).
    Raises DownloadError on any failure (no head response, stall, CRC mismatch).
    """
    # Reset voice buffer
    client.voice_data = bytearray()
    client.voice_packets = 0
    client.receiving = True

    # file_size from device = raw frame bytes; actual BLE data includes 9-byte headers
    # per 80-byte frame, so expected_size ≈ file_size * 89/80
    expected_size = int(file_size * 89 / 80)

    try:
        # SYNC_FILE_START (cmd 28): session_id, offset=0, file_size
        payload = struct.pack("<III", session_id, 0, file_size)
        await client.send(CMD_SYNC_FILE_START, payload)

        head = await client.wait_response(CMD_SYNC_FILE_START, timeout=10.0)
        if head is None:
            raise DownloadError("No file head response from device")

        if len(head) >= 5 and head[4] != 0:
            raise DownloadError(f"Transfer rejected by device (status={head[4]})")

        # Collect voice packets until SYNC_FILE_TAIL
        start = time.time()
        last_size = 0
        stall_count = 0
        tail = None

        while True:
            tail = await client.wait_response(CMD_SYNC_FILE_TAIL, timeout=0.5)
            if tail is not None:
                break

            current = len(client.voice_data)
            if current != last_size:
                stall_count = 0
                last_size = current
                if verbose:
                    elapsed = time.time() - start
                    speed = current / elapsed if elapsed > 0 else 0
                    pct = min(current / expected_size * 100, 100.0)
                    print(f"\r  {current}/{expected_size} ({pct:.1f}%) "
                          f"{speed / 1024:.1f} KB/s [{client.voice_packets} pkts]",
                          end="", flush=True)
            else:
                stall_count += 1
                if stall_count > 20:  # 10s with no data
                    if verbose:
                        print()
                    raise DownloadError(
                        f"Transfer stalled at {current}/{expected_size} bytes "
                        f"after {time.time() - start:.0f}s"
                    )

        if verbose:
            print()

        file_data = bytes(client.voice_data)

        if not file_data:
            raise DownloadError("Transfer completed but no data received")

        # Size sanity check: expected_size is approximate (89/80 expansion), so
        # allow 10% under-delivery to account for device rounding and padding.
        # Anything below 90% of expected indicates a truncated transfer.
        min_acceptable = int(expected_size * 0.90)
        if len(file_data) < min_acceptable:
            raise DownloadError(
                f"Transfer too short: got {len(file_data)} bytes, "
                f"expected >= {min_acceptable} (~{expected_size})"
            )

        # Structural invariant: BLE voice packets are 89 bytes each.
        # Non-aligned data indicates corruption, truncation, or protocol drift.
        if len(file_data) % 89 != 0:
            raise DownloadError(
                f"Data not 89-byte aligned: {len(file_data)} bytes "
                f"(remainder={len(file_data) % 89}). "
                f"Possible packet corruption or firmware protocol change."
            )

        # Tail packet validation — must contain enough bytes for CRC field
        if tail is None or len(tail) < 6:
            raise DownloadError(
                f"Malformed tail packet: expected >=6 bytes, "
                f"got {len(tail) if tail else 0}"
            )

        device_crc = struct.unpack("<H", tail[4:6])[0]
        local_crc = crc16_ccitt(file_data)

        # Device sends 0xFFFF when it skips CRC (voice-mode transfers)
        crc_skipped = device_crc == 0xFFFF
        if verbose:
            if crc_skipped:
                print(f"  CRC: skipped by device (0xFFFF)")
            else:
                match = "OK" if device_crc == local_crc else "MISMATCH"
                print(f"  CRC: 0x{device_crc:04x}/0x{local_crc:04x} {match}")

        await client.send(CMD_FILE_CHECKSUM, struct.pack("<BH", 0, local_crc))
        ack = await client.wait_response(CMD_FILE_CHECKSUM_RSP, timeout=5.0)
        if ack is None:
            raise DownloadError("No checksum acknowledgement from device")

        if not crc_skipped and device_crc != local_crc:
            raise DownloadError(
                f"CRC mismatch: device=0x{device_crc:04x} local=0x{local_crc:04x}"
            )

        return file_data

    finally:
        client.receiving = False

File: Sources/BLEKit/Transfer.swift
/// File download over BLE — voice packet capture approach.
///
/// Ported from Python CLI `src/plaude/ble/transfer.py`.

import Foundation

/// Raised when a file download fails.
public struct DownloadError: Error, LocalizedError {
    public let message: String
    public var errorDescription: String? { message }

    public init(_ message: String) {
        self.message = message
    }
}

/// Download a recording from the device.
///
/// The device sends file data as proto_type=0x02 (voice) packets after
/// SYNC_FILE_START, then sends SYNC_FILE_TAIL with CRC when done.
///
/// Returns raw Opus bytes (with 9-byte per-frame headers).
public func downloadFile(
    client: PlaudClient,
    sessionID: UInt32,
    fileSize: UInt32,
    verbose: Bool = false,
    progress: ((Int, Int, Double) -> Void)? = nil
) async throws -> Data {
    await client.resetVoiceBuffer()
    await client.setReceiving(true)

    // file_size from device = raw frame bytes; actual BLE data includes 9-byte headers
    // per 80-byte frame, so expected_size ≈ file_size * 89/80
    let expectedSize = Int(Double(fileSize) * 89.0 / 80.0)

    defer { Task { await client.setReceiving(false) } }

    // SYNC_FILE_START (cmd 28): session_id, offset=0, file_size
    var payload = Data(capacity: 12)
    var sid = sessionID.littleEndian
    var offset: UInt32 = 0
    var fs = fileSize.littleEndian
    payload.append(Data(bytes: &sid, count: 4))
    payload.append(Data(bytes: &offset, count: 4))
    payload.append(Data(bytes: &fs, count: 4))

    try await client.send(cmdSyncFileStart, payload: payload)

    guard let head = await client.waitResponse(cmdSyncFileStart, timeout: 10.0) else {
        throw DownloadError("No file head response from device")
    }

    if head.count >= 5 && head[4] != 0 {
        throw DownloadError("Transfer rejected by device (status=\(head[4]))")
    }

    // Collect voice packets until SYNC_FILE_TAIL
    let start = Date()
    var lastSize = 0
    var stallCount = 0

    while true {
        if let tail = await client.waitResponse(cmdSyncFileTail, timeout: 0.5) {
            // Got tail — validate and return
            let fileData = await client.voiceData

            if fileData.isEmpty {
                throw DownloadError("Transfer completed but no data received")
            }

            // Size sanity: allow 10% under-delivery
            let minAcceptable = Int(Double(expectedSize) * 0.90)
            if fileData.count < minAcceptable {
                throw DownloadError(
                    "Transfer too short: got \(fileData.count) bytes, expected >= \(minAcceptable) (~\(expectedSize))"
                )
            }

            // Structural invariant: BLE voice packets are 89 bytes each
            if fileData.count % 89 != 0 {
                throw DownloadError(
                    "Data not 89-byte aligned: \(fileData.count) bytes (remainder=\(fileData.count % 89))"
                )
            }

            // Tail packet validation
            guard tail.count >= 6 else {
                throw DownloadError("Malformed tail packet: expected >=6 bytes, got \(tail.count)")
            }

            let deviceCRC = tail.withUnsafeBytes { ptr -> UInt16 in
                ptr.loadUnaligned(fromByteOffset: 4, as: UInt16.self).littleEndian
            }
            let localCRC = crc16CCITT(fileData)
            let crcSkipped = deviceCRC == 0xFFFF

            if verbose {
                if crcSkipped {
                    print("  CRC: skipped by device (0xFFFF)")
                } else {
                    let match = deviceCRC == localCRC ? "OK" : "MISMATCH"
                    print("  CRC: 0x\(String(format: "%04x", deviceCRC))/0x\(String(format: "%04x", localCRC)) \(match)")
                }
            }

            // Send checksum acknowledgement
            var ackPayload = Data(capacity: 3)
            ackPayload.append(0)
            var crcLE = localCRC.littleEndian
            ackPayload.append(Data(bytes: &crcLE, count: 2))
            try await client.send(cmdFileChecksum, payload: ackPayload)
            _ = await client.waitResponse(cmdFileChecksumRsp, timeout: 5.0)

            if !crcSkipped && deviceCRC != localCRC {
                throw DownloadError(
                    "CRC mismatch: device=0x\(String(format: "%04x", deviceCRC)) local=0x\(String(format: "%04x", localCRC))"
                )
            }

            return fileData
        }

        // Check for progress / stall
        let current = await client.voiceData.count
        if current != lastSize {
            stallCount = 0
            lastSize = current
            let elapsed = Date().timeIntervalSince(start)
            let pct = min(Double(current) / Double(expectedSize) * 100.0, 100.0)
            progress?(current, expectedSize, pct)

            if verbose {
                let speed = elapsed > 0 ? Double(current) / elapsed : 0
                let pkts = await client.voicePacketCount
                print("\r  \(current)/\(expectedSize) (\(String(format: "%.1f", pct))%) \(String(format: "%.1f", speed / 1024)) KB/s [\(pkts) pkts]", terminator: "")
                fflush(stdout)
            }
        } else {
            stallCount += 1
            if stallCount > 20 {  // 10s with no data
                throw DownloadError(
                    "Transfer stalled at \(current)/\(expectedSize) bytes after \(String(format: "%.0f", Date().timeIntervalSince(start)))s"
                )
            }
        }
    }
}

